Assets/Scripts/Gameplay/ConnectionManagement/ConnectionState/OfflineState.cs

```diff
@@ -0,0 +1,197 @@
+                var utp = (UnityTransport)m_ConnectionManager.NetworkManager.NetworkConfig.NetworkTransport;
+                utp.SetHostRelayData(ipv4Address, port, allocationIdBytes, key, connectionData, isSecure: true);
+            }
+            catch (Exception e)
+            {
+                Debug.LogErrorFormat($"{e.Message}");
+                throw;
+            }
+
+            StartHost(GetPlayerId(), playerName);
+        }
+
+        public override void ApprovalCheck(byte[] connectionData, ulong clientId, NetworkManager.ConnectionApprovedDelegate connectionApprovedCallback)
+        {
+            // This happens when starting as a host, before the end of the StartHost call. In that case, we simply approve ourselves.
+            if (m_ConnectionManager.NetworkManager.IsHost && clientId == m_ConnectionManager.NetworkManager.LocalClientId)
+            {
+                var payload = System.Text.Encoding.UTF8.GetString(connectionData);
+                var connectionPayload = JsonUtility.FromJson<ConnectionPayload>(payload); // https://docs.unity3d.com/2020.2/Documentation/Manual/JSONSerialization.html
+
+                SessionManager<SessionPlayerData>.Instance.SetupConnectingPlayerSessionData(clientId, connectionPayload.playerId,
+                    new SessionPlayerData(clientId, connectionPayload.playerName, new NetworkGuid(), 0, true));
+
+                // connection approval will create a player object for you
+                connectionApprovedCallback(true, null, true, Vector3.zero, Quaternion.identity);
+            }
+        }
+
+        void SetConnectionPayload(string playerId, string playerName)
+        {
+            var payload = JsonUtility.ToJson(new ConnectionPayload()
+            {
+                playerId = playerId,
+                playerName = playerName,
+                isDebug = Debug.isDebugBuild
+            });
+
+            var payloadBytes = System.Text.Encoding.UTF8.GetBytes(payload);
+
+            m_ConnectionManager.NetworkManager.NetworkConfig.ConnectionData = payloadBytes;
+        }
+
+        void ConnectClient(string playerId, string playerName)
+        {
+            SetConnectionPayload(playerId, playerName);
+
+            if (m_ConnectionManager.NetworkManager.StartClient())
+            {
+                m_ConnectionManager.ChangeState(ClientConnecting);
+            }
+            else
+            {
+                m_ConnectStatusPublisher.Publish(ConnectStatus.StartClientFailed);
+            }
+        }
+
+        bool StartHost(string playerId, string playerName)
+        {
+            SetConnectionPayload(playerId, playerName);
+            var success = m_ConnectionManager.NetworkManager.StartHost();
+            if (!success)
+            {
+                m_ConnectStatusPublisher.Publish(ConnectStatus.StartHostFailed);
+            }
+            else
+            {
+                m_ConnectionManager.ChangeState(Hosting);
```

Comment:
I wouldn't hide state changing in a "StartHost". This seems like hidden behaviour. When reading "StartHost" from the callsite, I'd expect it to do some NGO things, nothing else.

Reply:
Good point! I moved it here to reduce code duplication but I agree that it makes more sense to not hide it

Reply:
The addition of the StartingHost state fixes this as well